### DISCUSSION TITLE
Estimate gas failed, simulate failed, but tx succeeded

### DISCUSSION THREAD

**Discussion by hungdoansy** [wevm/wagmi#4164] [Q&A]
Hi guys, I ran into this issue several times and now I know why

TLDR; you need to provide the `account` param when simulating/estimating gas

```ts
const params = {
    abi: erc20Abi,
    address: TOKEN_ADDRESS,
    functionName: "transfer",
    args: [
        // arguments,
    ],
    account: accountAddress, // this is VERY important, MUST add
} as const

const gas = await publicClient.estimateContractGas(params)

const { request } = await publicClient.simulateContract({
    ...params,
    // gas: calculateGasMargin(gas),
})

const hash = await writeContractAsync(params)
```

This is the code that I often use when I need to make a tx. I forgot to add `account` and it failed. Now I know why

**Comment by tmm** [comment]
FYI `account` is not required and falls back to [connected account](https://wagmi.sh/react/api/hooks/useAccount#address) when not provided.

**Reply by hungdoansy** [reply]
@tmm how sure are you about this? because it happened to me and I am 100% sure account did not automatically fall back to the connect account. After I manually added acount and it worked.

**Reply by tmm** [reply]
Might be some confusion here. Wagmi's [`estimateGas`](https://github.com/wevm/wagmi/blob/531ec22e4ecb7891a3e83ecf8d3b69b8879d3a5c/packages/core/src/actions/estimateGas.ts#L67) and [`simulateContract`](https://github.com/wevm/wagmi/blob/531ec22e4ecb7891a3e83ecf8d3b69b8879d3a5c/packages/core/src/actions/simulateContract.ts#L148) both fall back to the connected account while Viem's related methods do not (since they don't know what the connected account is as that is a Wagmi concern).
